### ISSUE TITLE
get citation feature

### ISSUE THREAD

**Issue by hongsudt** [informatics-isi-edu/ermrestjs#678]
We should provide a menu option for users to get citation associated with each record.. 

A good example can be found here: 
https://academic.oup.com/database/article/doi/10.1093/database/bav010/2433142

Edit:
To support this feature, we first need to define a table level annotation that `ermrestJS` can understand. This will look something like the following (based on bibtex terms):
```js
// Fields marked * are required

tag:isrd.isi.edu,2018:citation: {
  template_engine: "handlebars" or "mustache",
  author_pattern:  pattern expansion,
  title_pattern:   pattern expansion,
 *journal_pattern: pattern expansion,
 *year_pattern:    pattern expansion,
 *url_pattern:     pattern expansion,
  id_pattern:      pattern expansion
}
```

Example
```
{
  "template_engine": "handlebars",
  "author_pattern": "{{{$fkeys.Common.Collection_Principal_Investigator_fkey.values.Full_Name}}}",
   "journal_pattern": "GUDMAP Consortium",
   "year_pattern": "{{formatDate RCT 'YYYY'}}",
   "url_pattern": "{{#if Persistent_ID}} {{{Persistent_ID}}} {{else}} https://www.gudmap.org/id/{{{RID}}} {{/if}}",
   "id_pattern": "Collection_{{{RID}}}"
}
```
Should render to:

**Shared Link:**
[https://doi.org/123.4/A-BCDE](https://doi.org/123.4/A-BCDE)

**Citation:**
Carl Kesselman, Experiment Protocol, GUDMAP Consortium, [https://doi.org/123.4/A-BCDE](https://doi.org/123.4/A-BCDE) (2018).

**Download Citation File:**
BibTex

A new handlebars function needs to be introduced which will allow data modelers to take a date or timestamp[tz] and format the output for the "year" pattern appropriately. Generally only the year will be used for that value but some data modelers may want to be more specific.

Only the presence of the required fields will be validated. If any of the values are not the proper type, they will still be displayed in the citation.

**Comment by carlkesselman** [comment]
Would we expect to download the citation, or save it in the cutbuffer or something like that.

Also, I think it would be a UI nightmare to be able to generate a citation for something that is not dereferanceable via the resolver, so this again goes toward consolidating configuration for resolver into the catalog.

Carl
Sent from my iPad

----------------------------------------------------------
Dr. Carl Kesselman
Dean’s Professor, Epstein Department of Industrial and Systems Engineering
Fellow, Information Sciences Institute
Viterbi School of Engineering

Professor, Preventive Medicine
Keck School of Medicine

University of Southern California
4676 Admiralty Way, Suite 1001, Marina del Rey, CA <x-apple-data-detectors://1/0>  90292-6695<x-apple-data-detectors://1/0>
Phone:  +1 (310) 448-9338<tel:+1%20(310)%20448-9338>
Email:  carl@isi.edu<mailto:carl@isi.edu>
Web: http://www.isi.edu/~carl


On Jul 23, 2018, at 3:52 PM, Hongsuda <notifications@github.com<mailto:notifications@github.com>> wrote:


We should provide a menu option for users to get citation associated with each record..

A good example can be found here:
https://academic.oup.com/database/article/doi/10.1093/database/bav010/2433142

**Comment by hongsudt** [comment]
We will need to define a citation annoatation: 
```
   tag:misd.isi.edu,2018:citation = <citation object> 
Where <citation_object> consists of the following fields (based on bibtex terms): 
  * author
  * title 
  * journal 
  * year
  * doi 
  * url
```
Example
```
{
  author : {{{$fkeys.Protocol.Protocol_Principal_Investigator_fkey.values.Full_Name}}},
  journal: {{{$fkeys.Protocol.Protocol_Consortium_fkey.values.Name}}},
  title: "Experiment Protocol",
  doi:{{{Permanent_ID}}}, 
  URL:{{{Permanent_ID}}} or www.gudmap.org/id/{{{RID}}}
}
```
Should render to: 
Carl Kesselman, Experiment Protocol, GUDMAP consortium, https://doi.org/123.4/A-BCDE. 

@carlkesselman @karlcz @robes @ljpearlman what do you think?

**Comment by hongsudt** [comment]
Example of a bibtex
```
@article{doi:10.1093/database/bav010,
author = {Malladi, Venkat S. and Erickson, Drew T. and Podduturi, Nikhil R. and Rowe, Laurence D. and Chan, Esther T. and Davidson, Jean M. and Hitz, Benjamin C. and Ho, Marcus and Lee, Brian T. and Miyasato, Stuart and Roe, Gregory R. and Simison, Matt and Sloan, Cricket A. and Strattan, J. Seth and Tanaka, Forrest and Kent, W. James and Cherry, J. Michael and Hong, Eurie L.},
title = {Ontology application and use at the ENCODE DCC},
journal = {Database},
year = {2015},
doi = {10.1093/database/bav010},
URL = {http://dx.doi.org/10.1093/database/bav010},
}
```

**Comment by carlkesselman** [comment]
It would seem that some of this wouldn’t make sense if I am generating a citation to just some row in a table, would it?  Author, year (actually timestamp) and URL would make sense, others probably not so much so?

Carl

----------------------------------------------------------
Dr. Carl Kesselman
Dean’s Professor, Epstein Department of Industrial and Systems Engineering
Fellow, Information Sciences Institute
Viterbi School of Engineering

Professor, Preventive Medicine
Keck School of Medicine

University of Southern California
4676 Admiralty Way, Suite 1001, Marina del Rey, CA  90292-6695
Phone:  +1 (310) 448-9338
Email:  carl@isi.edu<mailto:carl@isi.edu>
Web: http://www.isi.edu/~carl








On Aug 22, 2018, at 3:32 PM, Hongsuda <notifications@github.com<mailto:notifications@github.com>> wrote:


We will need to define a citation annoatation:

   tag:misd.isi.edu<http://misd.isi.edu>,2018:citation = <citation object>
Where <citation_object> consists of the following fields (based on bibtex terms):
  * author
  * title
  * journal
  * year
  * doi
  * url


Example

{
  author : {{{$fkeys.Protocol.Protocol_Principal_Investigator_fkey.values.Full_Name}}},
  journal: {{{$fkeys.Protocol.Protocol_Consortium_fkey.values.Name}}},
  title: "Experiment Protocol",
  doi:{{{Permanent_ID}}},
  URL:{{{Permanent_ID}}} or www.gudmap.org/id/<http://www.gudmap.org/id/>{{{RID}}}
}


Should render to:
Carl Kesselman, Experiment Protocol, GUDMAP consortium, https://doi.org/123.4/A-BCDE.

@carlkesselman<https://github.com/carlkesselman> @karlcz<https://github.com/karlcz> @robes<https://github.com/robes> @ljpearlman<https://github.com/ljpearlman> what do you think?

**Comment by hongsudt** [comment]
Other fields are optional. For GUDMAP, we put the PI as the author, and likely to leave the title blank..

**Comment by carlkesselman** [comment]
This is orthogonal I think.  We need to be building this from the perspective that these are light-weight citations that should be super easy to create and used pretty much anywhere/everywhere.  We should not be building in more complex metadata rules like who is the author.  Author is creator of the record, creation time is in the record.  Name could be a name field if it exists, etc.

If we really want to create a citation that would be referred to via say bibtex, we would go through a more structured publication process.

Carl

----------------------------------------------------------
Dr. Carl Kesselman
Dean’s Professor, Epstein Department of Industrial and Systems Engineering
Fellow, Information Sciences Institute
Viterbi School of Engineering

Professor, Preventive Medicine
Keck School of Medicine

University of Southern California
4676 Admiralty Way, Suite 1001, Marina del Rey, CA  90292-6695
Phone:  +1 (310) 448-9338
Email:  carl@isi.edu<mailto:carl@isi.edu>
Web: http://www.isi.edu/~carl








On Aug 22, 2018, at 4:25 PM, Hongsuda <notifications@github.com<mailto:notifications@github.com>> wrote:

Not necessary.. In some project, you might not want to disclose the owner. Also, the person that creates the record might not be the owner of the records e.g. we created many records for members.. And RCB is nullable.

Until the policy regarding the resolver is changed, unless there is an annotation, there won't be a citation.


________________________________

**Comment by carlkesselman** [comment]
I was thinking super simple here… we package up the RID, wrap it in the URI prefix (gudmap.org/id/XXX-XXX<http://gudmap.org/id/XXX-XXX>) and you’ve got your citation.  If we wanted to have an alternative where we download a record with some other data, ok.  Lets think about how we are going to exchange these…. text messages, slack, email, slap them into a google doc, stick in your powerpoint deck…..

Carl

----------------------------------------------------------
Dr. Carl Kesselman
Dean’s Professor, Epstein Department of Industrial and Systems Engineering
Fellow, Information Sciences Institute
Viterbi School of Engineering

Professor, Preventive Medicine
Keck School of Medicine

University of Southern California
4676 Admiralty Way, Suite 1001, Marina del Rey, CA  90292-6695
Phone:  +1 (310) 448-9338
Email:  carl@isi.edu<mailto:carl@isi.edu>
Web: http://www.isi.edu/~carl








On Aug 22, 2018, at 4:25 PM, Hongsuda <notifications@github.com<mailto:notifications@github.com>> wrote:

Not necessary.. In some project, you might not want to disclose the owner. Also, the person that creates the record might not be the owner of the records e.g. we created many records for members.. And RCB is nullable.

Until the policy regarding the resolver is changed, unless there is an annotation, there won't be a citation.


________________________________

**Comment by jrchudy** [comment]
@hongsudt  support for this feature has been added to both `ermrestJS` and `chaise`. Documentation for the annotation can be found [here](https://github.com/informatics-isi-edu/ermrest/blob/master/docs/user-doc/annotation.md#2018-citation).

Reassigning this issue to you for verification.

**Comment by carlkesselman** [comment]
Great thanks.  Overall this looks nice.

Of course, we want the RID resolver link, not the chaise link (I suspect you are waiting for the resolver?) We should think a bit about the text.  We have Share Citation, Share Link and Citation.  Seems a bit confusing to me.

Carl

----------------------------------------------------------
Dr. Carl Kesselman
Dean’s Professor, Epstein Department of Industrial and Systems Engineering
Fellow, Information Sciences Institute
Viterbi School of Engineering

Professor, Preventive Medicine
Keck School of Medicine

University of Southern California
4676 Admiralty Way, Suite 1001, Marina del Rey, CA  90292-6695
Phone:  +1 (310) 448-9338
Email:  carl@isi.edu<mailto:carl@isi.edu>
Web: http://www.isi.edu/~carl








On Sep 13, 2018, at 5:22 PM, Hongsuda <notifications@github.com<mailto:notifications@github.com>> wrote:


Check this link for citation with DOI url.
https://dev.rebuildingakidney.org/chaise/recordset/#2/Common:Collection@sort(RID)

**Comment by carlkesselman** [comment]
Is there a way to get a “copy link” action there somehow?

Carl

----------------------------------------------------------
Dr. Carl Kesselman
Dean’s Professor, Epstein Department of Industrial and Systems Engineering
Fellow, Information Sciences Institute
Viterbi School of Engineering

Professor, Preventive Medicine
Keck School of Medicine

University of Southern California
4676 Admiralty Way, Suite 1001, Marina del Rey, CA  90292-6695
Phone:  +1 (310) 448-9338
Email:  carl@isi.edu<mailto:carl@isi.edu>
Web: http://www.isi.edu/~carl








On Sep 13, 2018, at 5:22 PM, Hongsuda <notifications@github.com<mailto:notifications@github.com>> wrote:


Check this link for citation with DOI url.
https://dev.rebuildingakidney.org/chaise/recordset/#2/Common:Collection@sort(RID)

**Comment by hongsudt** [comment]
Isn't that the same as right click on link?

**Comment by hongsudt** [comment]
The title will be just "Share" and there will be a bit improvement on the UX. Please check the UX again after the code is merged.

**Comment by carlkesselman** [comment]
yea, I suppose so.

Google drive has this nice feature that they provide a button that copies the link to the clipboard.

[cid:3A2DA36B-BB55-483C-A9B3-A1F47438BA55]
----------------------------------------------------------
Dr. Carl Kesselman
Dean’s Professor, Epstein Department of Industrial and Systems Engineering
Fellow, Information Sciences Institute
Viterbi School of Engineering

Professor, Preventive Medicine
Keck School of Medicine

University of Southern California
4676 Admiralty Way, Suite 1001, Marina del Rey, CA  90292-6695
Phone:  +1 (310) 448-9338
Email:  carl@isi.edu<mailto:carl@isi.edu>
Web: http://www.isi.edu/~carl








On Sep 13, 2018, at 5:51 PM, Hongsuda <notifications@github.com<mailto:notifications@github.com>> wrote:


Isn't that the same as right click on link?

**Comment by jrchudy** [comment]
The text and CSS changes have been merged as part of PR [#1643 in chaise](https://github.com/informatics-isi-edu/chaise/pull/1643).

**Comment by hongsudt** [comment]
I have verified that this works on rbk. SO I am closing this issue.
